feat: Ability to delete a contact

aswindevps opened the pull request:
# Pull Request Template

## Description

This change allows the administrator user to delete a contact and its related data like conversations, contact inboxes, and reports.

Fixes #1929

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

1. Deleting the contact should clear the contact and all associated conversations
2. Only Admin users should be able to delete a contact
3. Agents shouldn't be able to see the delete contact option
4. Agents shouldn't be able to hit delete contact endpoint
5. Reports should reflect correct data concerning the conversations that were deleted because of contact delete


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

aswindevps commented:
![Screenshot 2021-09-10 at 10 40 07 AM](https://user-images.githubusercontent.com/5848740/132802894-43b377e7-4d47-40ca-9c55-8c3d08b4d56b.png)
![Screenshot 2021-09-10 at 10 41 00 AM](https://user-images.githubusercontent.com/5848740/132802956-e0ed82a4-66e0-4f5f-a6c2-97c2de67aba2.png)

muhsin-k commented on app/controllers/api/v1/accounts/contacts_controller.rb:
```diff
@@ -70,6 +70,18 @@ def update
+    if ::OnlineStatusTracker.get_presence(
+      @contact.account.id, 'Contact', @contact.id
+    )
+      return render_error({ message: "#{@contact.name.capitalize} is Online, please try again later" },
```
**I18n**

aswindevps replied:
Added I18n

nithindavid commented on app/javascript/dashboard/routes/dashboard/conversation/contact/ContactInfo.vue:
```diff
@@ -80,11 +109,24 @@
+      :confirm-value="contact.name"
+      :confirm-place-holder-text="confirmPlaceHolderText"
+      @on-confirm="confirmDeletion"
+      @on-close="closeDelete"
```
Since @ already signifies an event we can rename this to `@close`

nithindavid replied:
Same for on-confirm to @confirm

aswindevps replied:
`woot-confirm-delete-modal` is common component. if we change @on-confirm to @confirm there are 4 places that need to be changed. 

Can we take this as a separate PR?

nithindavid commented on app/javascript/dashboard/routes/dashboard/contacts/components/ContactInfoPanel.vue:
```diff
@@ -3,7 +3,7 @@
       <i class="ion-android-close close-icon" />
     </span>
-    <contact-info show-new-message :contact="contact" />
+    <contact-info show-new-message :contact="contact" @onPanelClose="onClose" />
```
Can we use camel case for @onPanelClose ? Or simply rename to => @close or @panel-close

aswindevps replied:
Renamed to `@panel-close`

nithindavid commented on app/javascript/dashboard/helper/actionCable.js:
```diff
@@ -115,6 +116,11 @@ class ActionCableConnector extends BaseActionCableConnector {
   };
+
+  onContactDelete = data => {
+    this.app.$store.dispatch('contacts/deleteContactConversations', data.id);
```
```suggestion
    this.app.$store.dispatch('contacts/deleteContactThroughConversations', data.id);
```

aswindevps replied:
Updated suggestion

nithindavid reviewed (changes requested):
Have left some minor comments.

aswindevps commented:
@pranavrajs @nithindavid @sojan-official 
Completed testing for all channels except Twitter(Not able to add twitter channel in staging. already discussed with @sojan-official). Everything looks good. 

Addressed all the review comments. Please let me know if any other changes required.
